_posts/2017-09-28-presentation-grpc.md

```diff
@@ -0,0 +1,332 @@
+
+Chaque service rpc est déclaré dans le fichier protobuf.
+
+gRPC permet quatre modes communications.
+
+
+Le one to one classique :
+```
+Client Request -> Server
+Client <- Response Server
+
+service CustomService {
+    rpc Endpoint(Request) returns (Response) {}
+}
+```
+
+La streaming côté client :
+```
+Streaming Request -> Server
+Client <- Response Server
+
+service CustomService {
+    rpc Endpoint(stream Request) returns (Response) {}
+}
+```
+
+Le streaming côté serveur :
+
+```
+Request -> Server
+Client <- Streaming Response
+
+service CustomService {
+    rpc Endpoint(Request) returns (stream Response) {}
+}
+```
+
+Le streaming bidirectionnel :
+```
+Streaming Request -> Server
+Client -> Streaming Response
+
+service CustomService {
+    rpc Endpoint(stream Request) returns (stream Response) {}
+}
+```
+
+Nous allons voir pour cet article uniquement le troisième mode `streaming côté serveur` mais les quatre sont très semblables.
+
+Vous pouvez retrouver l'ensemble du code de [l'exemple sur mon github](https://github.com/qneyrat/go-grpc-example).
+
+### Installation
+
+Assurez vous avant de commencer l'installation d'avoir bien installé Go en version supérieur à 1.5 et Protocol Buffers en version 3.
+
+Récupérez gRPC pour Go :
+
+```Bash
+go get google.golang.org/grpc
+```
+
+### Protobuf
+
+Nous allons commencer par récupérer le fichier proto du [précédent article](https://blog.eleven-labs.com/fr/presentation-protocol-buffers/).
+
+Nous allons ajouté un service pour récupérer en streaming la liste des Posts.
+
+Un service rpc est composé :
+
+```
+rpc function(request) returns (response)
+```
+
+Ici on a une request vide et un streaming de Post.
+
+```Proto
+service PostService {
+    rpc ListPosts(google.protobuf.Empty) returns (stream Post) {}
+}
+```
+
+Ce qui nous donne :
+
+```Proto
+syntax = "proto3";
+
+import "google/protobuf/empty.proto";
+
+package main;
+
+message Post {
+    int32 id = 1;
+    string title = 2;
+    string author = 3;
+}
+
+service PostService {
+    rpc ListPosts(google.protobuf.Empty) returns (stream Post) {}
+}
+
+```
+
+### Serveur
+
+On va commencé par générer le code source depuis le fichier protobuf.
+
+```Bash
+protoc --proto_path=. --go_out=plugins=grpc:. post.proto
+```
+
+On commence par déclarer une pseudo base de données :
+
+```Go
+var posts = []Post{
+	{
+		Id:     1,
+		Title:  "My awesome article 1",
+		Author: "Quentin Neyrat",
+	},
+	{
+		Id:     2,
+		Title:  "My awesome article 2",
+		Author: "Quentin Neyrat",
+	},
+	{
+		Id:     3,
+		Title:  "My awesome article 3",
+		Author: "Quentin Neyrat",
+	},
+}
+```
+
+Puis on crée un serveur TCP sur le port 4000 pour notre serveur gRPC et on attache notre service déclaré dans le protobuf :
+
+```Go
+	lis, _ := net.Listen("tcp", fmt.Sprintf("localhost:%d", 4000))
```

Comment:
Maybe you can avoid `Sprintf` here as you do not used it later in the post? It will be more simpler to use ;)

```
lis, _ := net.Listen("tcp", "localhost:4000")
```